presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPageSourceProvider.java

```diff
@@ -83,34 +78,104 @@ public ConnectorPageSource createPageSource(
+            implements ConnectorPageSource
     {
-        int[] positions = new int[page.getPositionCount()];
-        int length = 0;
-        for (int position = 0; position < page.getPositionCount(); ++position) {
-            if (positionMatchesPredicate(page, position, domains)) {
-                positions[length++] = position;
+        private final FixedPageSource delegate;
+        private final List<ColumnHandle> columns;
+        private final DynamicFilter dynamicFilter;
+        private final boolean enableLazyDynamicFiltering;
+
+        private DynamicFilteringPageSource(FixedPageSource delegate, List<ColumnHandle> columns, DynamicFilter dynamicFilter, boolean enableLazyDynamicFiltering)
+        {
+            this.delegate = delegate;
+            this.columns = columns;
+            this.dynamicFilter = dynamicFilter;
+            this.enableLazyDynamicFiltering = enableLazyDynamicFiltering;
+        }
+
+        @Override
+        public long getCompletedBytes()
+        {
+            return delegate.getCompletedBytes();
+        }
+
+        @Override
+        public long getReadTimeNanos()
+        {
+            return delegate.getReadTimeNanos();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return delegate.isFinished();
+        }
+
+        @Override
+        public Page getNextPage()
+        {
+            if (enableLazyDynamicFiltering && !dynamicFilter.isComplete()) {
+                return null;
+            }
+            TupleDomain<ColumnHandle> predicate = dynamicFilter.getCurrentPredicate();
+            if (predicate.isNone()) {
+                close();
+                return null;
             }
+            Page page = delegate.getNextPage();
+            if (page != null) {
+                page = applyFilter(page, predicate.transform(columns::indexOf).getDomains().get());
+            }
+            return page;
         }
-        return page.getPositions(positions, 0, length);
-    }
 
-    private boolean positionMatchesPredicate(Page page, int position, Map<Integer, Domain> domains)
-    {
-        for (Map.Entry<Integer, Domain> entry : domains.entrySet()) {
-            int channel = entry.getKey();
-            Domain domain = entry.getValue();
-            Object value = TypeUtils.readNativeValue(domain.getType(), page.getBlock(channel), position);
-            if (!domain.includesNullableValue(value)) {
-                return false;
+        @Override
+        public CompletableFuture<?> isBlocked()
+        {
+            return enableLazyDynamicFiltering ? dynamicFilter.isBlocked() : NOT_BLOCKED;
+        }
+
+        @Override
+        public long getSystemMemoryUsage()
+        {
+            return delegate.getSystemMemoryUsage();
+        }
+
+        @Override
+        public void close()
+        {
+            delegate.close();
+        }
+
+        private static Page applyFilter(Page page, Map<Integer, Domain> domains)
```

Comment:
move these to top-level class `MemoryPageSourceProvider`. This way there should be less changes overall